WIP bpo-34589: Move locale coercion back to the CLI app

ncoghlan opened the pull request:
This PR moves locale coercion back to happening as soon as possible in the CLI app, so there's no re-encoding dance involved in handling it, and so you can emulate systems that don't have a suitable target locale with `PYTHONCOERCECLOCALE=0`, even when working on code that runs with `-E` or `-I`.

To help with the documentation enhancements in the PR, it also implements [[bpo-30672](https://bugs.python.org/issue30672)](https://www.bugs.python.org/issue30672) (which makes the POSIX locale behave the same way as it does on glibc, even when it's not a simple alias for the default C locale)

<!-- issue-number: [[bpo-34589](https://bugs.python.org/issue34589)](https://www.bugs.python.org/issue34589) -->
https://bugs.python.org/issue34589
<!-- /issue-number -->

ncoghlan committed:
bpo-34589: Move locale coercion back to the CLI app

ncoghlan commented on Python/pylifecycle.c:
```diff
@@ -350,24 +350,25 @@ _coerce_default_locale_settings(int warn, const _LocaleCoercionTarget *target)
+                       const char **coercion_warning)
 {
+    int locale_was_coerced = 0;
+    *coercion_target = NULL;
```
This function is missing checks for `coercion_target` and `coercion_warning` being NULL, and skipping the assignments in that case.

ncoghlan replied:
I added these missing checks in https://github.com/python/cpython/pull/9257/commits/f4c686babca9870fc832797f2b960a6420c14f38

vstinner reviewed (dismissed):
I dislike the idea of having PYTHONCOERCECLOCALE env var which behaves differently than all other PYTHON* env vars. See my -X proposal: https://bugs.python.org/issue34589#msg325282

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-34589-move-locale-coercion-back-to-the-intended-location

ncoghlan committed:
Add NEWS entry

ncoghlan committed:
Update documentation

ncoghlan committed:
Add missing period

ncoghlan committed:
Fix option markup (courtesy of 'make suspicious')

ncoghlan committed:
Handle NULL output pointers in _Py_CoerceLegacyLocale

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-34589-move-locale-coercion-back-to-the-intended-location

ncoghlan committed:
Further adjustments to locale coercion

- PYTHONCOERCECLOCALE is handled the same way as LANG,
  LC_CTYPE and LC_ALL (i.e. independently of -E and -I)
- locale coercion tests are once again running in isolated
  mode
- be explicit that the env var key and value must be encoded
  as ASCII to have any effect
- implement bpo-30672, such that the POSIX locale is always
  handled the same way as the C locale, even when it isn't a
  simple platform level alias for the latter

ncoghlan committed:
bpo-30672: Make POSIX locale coercion tests unconditional

ericsnowcurrently commented on Modules/main.c:
```diff
@@ -1739,6 +1733,7 @@ pymain_init(_PyMain *pymain, PyInterpreterState **interp_p)
     pymain_clear_config(config);
     config = &interp->core_config;
 
+
```
accidental blank line?

ncoghlan replied:
Deleted

ericsnowcurrently commented on Modules/main.c:
```diff
@@ -1789,6 +1784,42 @@ _Py_UnixMain(int argc, char **argv)
+        */
+    _Py_SetLocaleFromEnv(LC_CTYPE);
+    if (_Py_LegacyLocaleDetected()) {
+            /* We ignore the Python -E and -I flags here, as the CLI needs to sort out
```
mis-indented?

ncoghlan replied:
Yep

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -273,29 +273,31 @@ int
     const char *ctype_loc = setlocale(LC_CTYPE, NULL);
-    return ctype_loc != NULL && strcmp(ctype_loc, "C") == 0;
+    return ctype_loc != NULL &&
+           (strcmp(ctype_loc, "C") == 0 || strcmp(ctype_loc, "C") == 0);
```
Did you mean for one of these two to be `"c"` (lower-case) or something?

ncoghlan replied:
Copy-and-paste error (second one was supposed to POSIX), combined with deferral of the related tests to a follow-up PR for https://www.bugs.python.org/issue30672

ericsnowcurrently commented on Modules/main.c:
```diff
@@ -1303,6 +1306,9 @@ pymain_read_conf(_PyMain *pymain, _PyCoreConfig *config,
     /* Set LC_CTYPE to the user preferred locale */
     _Py_SetLocaleFromEnv(LC_CTYPE);
 
+    /* TODO: With locale coercion moved back to _PyUnix_Main, this can
```
The function comment above for `pymain_read_conf()` is out-of-date with this change, no?

ncoghlan replied:
Clarified (also filed https://bugs.python.org/issue34914 which I was planning to do anyway, updating that comment just reminded me I hadn't done it yet)

ericsnowcurrently commented on Python/coreconfig.c:
```diff
@@ -762,6 +761,30 @@ get_env_flag(_PyCoreConfig *config, int *flag, const char *name)
+                } else {
+                    config->warn_on_c_locale = 0;
+                }
+            }
```
wrong indentation?

ncoghlan replied:
Fixed.

ericsnowcurrently commented on Python/pylifecycle.c:
```diff
@@ -344,24 +346,29 @@ _coerce_default_locale_settings(int warn, const _LocaleCoercionTarget *target)
-void
-_Py_CoerceLegacyLocale(int warn)
+int
+_Py_CoerceLegacyLocale(const char **coercion_target,
```
It may be worth noting somewhere that this is one of the few C-API functions that may be called before the runtime is configured (or possibly even initialized).  I know it's somewhat orthogonal to the objectives of this PR, but such a comment would give a little context.

ncoghlan replied:
Added a note to both this and the legacy locale detection API.

The downside of the notes I added is that a lot of other functions have those same restrictions, but *don't* have a marker comment like the ones I just added, so they may give folks an incorrect impression.

So it may be better to take those comments out again, and file an enhancement issue on BPO pointing out that this is a maintainability issue, where it's not obvious which parts of the code need to be careful about only calling pre-init friendly APIs.

ericsnowcurrently commented on Modules/main.c:
```diff
@@ -1729,6 +1717,12 @@ pymain_init(_PyMain *pymain, PyInterpreterState **interp_p)
     pymain_init_stdio(pymain, config);
 
+    if (config->warn_on_c_locale && pymain->locale_coercion_warning != NULL) {
+        fprintf(stderr, pymain->locale_coercion_warning, pymain->locale_coercion_target);
```
If we used sprintf in `_Py_CoerceLegacyLocale()` then we'd only need to return the fully-formed warning from it and we could drop `pymain->locale_coercion_target` entirely.  I mention this because that field was a little misleading/confusing at first (I was expecting it be used elsewhere).

Did you have some other plan for that field?

ncoghlan replied:
The issue I had with doing the message formatting in _Py_CoerceLegacyLocale is coping with the dynamic memory allocation needed (you could use a statically allocated array instead, but that's a different kind of ick).

However, I'm thinking there may be a better way to defer the rendering of the warning while still only needing one field on the main struct: make the field a function pointer accepting a C output stream that gets called instead of a C string that gets printed, and then in the warning rendering function, substitute the actual current locale into the message, rather than saving the name of the coercion target (This is all internal to the CPython command line app, so we don't need to worry about intervening locale changes).

With that change, this would just be a standard C callback API, rather than something that tightly couples the code that determines the warning is needed to the code that actually prints it after the standard streams are configured.

ericsnowcurrently commented on Misc/NEWS.d/next/Core and Builtins/2018-09-20-21-36-57.bpo-34589.9zERCD.rst:
```diff
@@ -0,0 +1,7 @@
+The locale coercion implementation has been adjusted back to being closer to
+the design documented in PEP 538. This means locale coercion is once again
+solely the responsibility of the application embedding the CPython runtime
```
I'm not sure I follow the distinction you are describing here.  As far as I can tell, the role of the embedding app has not changed at all.  By "CPython runtime" are you talking about the internal C-API?  Otherwise I'm just not seeing it.

Furthermore, from what I can tell, this change is really about 2 specific things:
* make `PYTHONCOERCECLOCALE` independent of the `-E` flag
* only use `PYTHONCOERCECLOCALE` if a legacy locale is detected

ncoghlan replied:
As released in 3.7.0, locale coercion happens at the point where `Py_Initialize` gets called (as that then internally coerces the locale to something else). That's appropriate for UTF-8 mode (since that only affects the CPython runtime in the current process, and the code carefully goes back and recalculates the settings that were already determined based on the locale encoding), but not for locale coercion (since that affects all subsequently executed locale-aware code in the current process, as well as in subprocesses, but can't reliably go back and rewrite all previously determined values in the embedding application).

That's not the way PEP 538 was originally implemented, and deliberately so: instead, the locale coercion happened in *the CLI app*, before the first call to `Py_DecodeLocale`, so nothing would ever get decoded as ASCII in the first place.

While that CLI app code has since moved to a private API inside the shared library (as `_Py_UnixMain`), this PR still gets it back to the same relative location: before the first call to `Py_DecodeLocale`, so its effectively indistinguishable from `LC_CTYPE=C.UTF-8` actually being set externally in the environment.

ncoghlan replied:
A potential analogy for why the way 3.7 is handling this is broken (and why PEP 538 deliberately defined it as working differently): performing locale coercion in `Py_Main` and `Py_Initialize` is akin to a Python library reconfiguring the logging module as a side effect of doing something else.

ncoghlan replied:
Amusingly enough, from the comments in the `readline` module implementation:

```
/* GNU readline() mistakenly sets the LC_CTYPE locale.
 * This is evil.  Only the user or the app's main() should do this!
 * We must save and restore the locale around the rl_initialize() call.
 */
```

ericsnowcurrently reviewed (changes requested):
The change makes sense to me, especially in light of the following from PEP 538:

> The interpreter will always check for the PYTHONCOERCECLOCALE environment variable at startup (even when running under the -E or -I switches), as the locale coercion check necessarily takes place before any command line argument processing. For consistency, the runtime check to determine whether or not to suppress the locale compatibility warning will be similarly independent of these settings.

There are a few minor things to address otherwise.

One point of clarification:  is it intentional that the warning is not emitted under `-E` even if the locale is coerced?

ncoghlan committed:
Address cosmetic review comments

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-34589-move-locale-coercion-back-to-the-intended-location

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-34589-move-locale-coercion-back-to-the-intended-location

ncoghlan committed:
Fix remaining merge issue

ncoghlan committed:
Note another potential test fragility resolution

ncoghlan committed:
Ensure there's no implicit C locale coercion

ncoghlan committed:
Reinstate support for -E/-I affecting PYTHONCOERCECLOCALE

ncoghlan commented on Python/coreconfig.c:
```diff
@@ -856,6 +855,30 @@ get_env_flag(_PyCoreConfig *config, int *flag, const char *name)
+                config->warn_on_c_locale = 1;
+            } else {
+                config->warn_on_c_locale = 0;
+            }
```
Note to self: still need to revert this.

ncoghlan commented on Python/pylifecycle.c:
```diff
@@ -262,30 +262,86 @@ initexternalimport(PyInterpreterState *interp)
+{
+    int enable_coercion = -1;
+    if (argc >= 1 && argv != NULL) {
+        /* Rudimentary arg parsing to allow -E and -I to disable
```
I was hoping to get away without needing to rely on the full `getopt` argument definition string here, but realised that's a futile endeavour due to the fact that cases like `python -iEV` are perfectly valid option strings.

ncoghlan replied:
Not so futile after all - with all the cross-platform getopt code migrated to `wchar_t`, it proved simpler to keep this early pass looking specifically for `-E` and `-I` independent of it.

Even adding a `-X no_locale_coercion` option would be straightforward (but I've chosen not to implement that at this stage)

ncoghlan committed:
Address self-review comments

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-34589-move-locale-coercion-back-to-the-intended-location
